router/throttler/factory_test.go

```diff
@@ -71,6 +74,27 @@ func TestFactory(t *testing.T) {
 			return ta.getLimit() == int64(0)
 		}, 2*time.Second, 100*time.Millisecond)
 	})
+
+	t.Run("check stats when adaptive is enabled", func(t *testing.T) {
+		config := config.New()
+		config.Set("Router.throttler.adaptive.enabled", true)
+		maxLimit := int64(300)
+		window := 2 * time.Second
+		config.Set("Router.throttler.adaptive.maxLimit", maxLimit)
+		config.Set("Router.throttler.adaptive.minLimit", int64(100))
+		config.Set("Router.throttler.adaptive.timeWindow", window)
+		ctrl := gomock.NewController(t)
+		mockStats := mock_stats.NewMockStats(ctrl)
+		mockMeasurement := mock_stats.NewMockMeasurement(ctrl)
```

Comment:
could [memstat](https://github.com/rudderlabs/rudder-go-kit/tree/main/stats/memstats) be more suitable than mock_stats?

Reply:
+1 for mem stats. One downside with the current mocks is we are not verifying the tags and well as which Gauge metrics is emitting what? With memstats, we would need to do it?

```GO
	t.Run("check stats when adaptive is enabled", func(t *testing.T) {
		maxLimit := int64(300)
		window := 2 * time.Second

		c := config.New()
		c.Set("Router.throttler.adaptive.enabled", true)
		c.Set("Router.throttler.adaptive.maxLimit", maxLimit)
		c.Set("Router.throttler.adaptive.minLimit", int64(100))
		c.Set("Router.throttler.adaptive.timeWindow", window)

		statsStore, err := memstats.New()
		require.NoError(t, err)

		f, err := NewFactory(c, statsStore)
		require.NoError(t, err)
		defer f.Shutdown()

		_ = f.Get("destName", "destID")

		require.EqualValues(t, maxLimit/getWindowInSecs(window), statsStore.Get("throttling_rate_limit", stats.Tags{
			"destinationId": "destID",
			"destType":      "destName",
			"adaptive":      "true",
		}).LastValue())
		require.EqualValues(t, 1, statsStore.Get("adaptive_throttler_limit_factor", stats.Tags{
			"destinationId": "destID",
			"destType":      "destName",
		}).LastValue())
	})

Reply:
Done